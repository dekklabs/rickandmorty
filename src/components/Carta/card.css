.carta {
    width: 240px;
    transition: all .3s ease-in;
}

.carta img {
    cursor: pointer;
}

.carta:hover {
    box-shadow: 1px 2px 14px 4px #ece6e6;
    transition: all .3s ease-in;
}

.vistaprevia {
    cursor: pointer;
}