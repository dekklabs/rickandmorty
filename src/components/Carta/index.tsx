import React, {useContext, useState} from 'react'
import './card.css'
import { 
    Card, 
    Button,
    Modal
} from "react-bootstrap";
import ModalContent from "../ModalContent";
import { DataContext } from "../../utils/DataContext";

const Carta = ({personaje:persona, setCart}:any) => {

    const {cart} = useContext(DataContext)

    const [show, setShow] = useState(false);

    const {id, name, image, location } = persona

    let price = 0
    if (name !== undefined) {
        price = name.charCodeAt(0)
    }

    const handleClose = () => setShow(false);
    const handleShow = () => setShow(true);

    const handleOnCart = (id) => {
        const existe = cart.some(producto => producto.id === id)

        if (existe) {
            setShowExiste(true);
        } else {
            setCart([
                ...cart,
                persona
            ])
        }
    }

    // Modal existe
    const [showExiste, setShowExiste] = useState(false);

    const handleCloseExiste = () => setShowExiste(false);
    const handleShowExiste = () => setShowExiste(true);

    return (
        <>
            <Card 
                className="carta"
            >
                <Card.Img variant="top" src={image} />
                <Card.Body>
                    <Card.Title>{name}</Card.Title>
                    <Card.Text>
                        {location?.description}
                    </Card.Text>
                    <Card.Text>
                        Price: S/ {price}
                    </Card.Text>
                    <div className="d-flex justify-content-between align-items-center mt-4">
                        <Card.Text 
                            className="mb-0 vistaprevia"
                            onClick={handleShow}
                        >
                            Vista previa
                        </Card.Text>
                        <Button 
                            variant="primary"
                            onClick={() => handleOnCart(id)}
                        >Agregar</Button>
                    </div>
                </Card.Body>
            </Card>
            
            <Modal 
                show={show} 
                onHide={handleClose}
            >
                <ModalContent
                    data={persona}
                    setCart={setCart}
                />
                <Modal.Footer>
                    <Button variant="danger" onClick={handleClose}>
                        Cerrar
                    </Button>
                </Modal.Footer>
            </Modal>

            {/* SI un item ya existe en el carrito */}
            <Modal 
                show={showExiste} 
                onHide={handleCloseExiste}
            >
                <Modal.Body>
                    <div className="text-center">
                        <h2>No se puede agregar dos veces el mismo producto :)</h2>
                        <Button variant="danger" onClick={handleCloseExiste}>
                            Cerrar
                        </Button>
                    </div>
                </Modal.Body>
            </Modal>
        </>
    )
}

export default Carta