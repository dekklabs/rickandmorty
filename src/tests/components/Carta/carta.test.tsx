import React from 'react';
import '@testing-library/jest-dom';
import { shallow } from 'enzyme'
import Carta from '../../../components/Carta'

describe('Pruebas en Card component', () => {

    const setCart = jest.fn();

    const persona = [
        {
            "id": 1,
            "name": "Rick Sanchez",
            "status": "Alive",
            "species": "Human",
            "type": "",
            "gender": "Male",
            "origin": {
              "name": "Earth (C-137)",
              "url": "https://rickandmortyapi.com/api/location/1"
            },
            "location": {
              "name": "Earth (Replacement Dimension)",
              "url": "https://rickandmortyapi.com/api/location/20"
            },
            "image": "https://rickandmortyapi.com/api/character/avatar/1.jpeg",
            "episode": [
              "https://rickandmortyapi.com/api/episode/1",
            ],
            "url": "https://rickandmortyapi.com/api/character/1",
            "created": "2017-11-04T18:48:46.250Z"
        }
    ]

    let wrapper = shallow(<Carta 
        personaje={persona}
        setCart={setCart}
    />)

    test('Debe sacar mostrar', () => {
        expect(wrapper).toMatchSnapshot()
    })
})