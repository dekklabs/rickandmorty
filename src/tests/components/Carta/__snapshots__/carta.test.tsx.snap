// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en Card component Debe sacar mostrar 1`] = `ShallowWrapper {}`;
